src/alg/dpll.py

```diff
@@ -0,0 +1,103 @@
+#######################################################################################################
+
+# simplify the cnf by assigning one of the variable to true or false and deleating clauses and literals 
+def conditioning(cnf,var):
+    new_cnf = list()
+    for entry in cnf:
+        if (len(entry)==0):
+            new_cnf.append(np.array([]))
+        elif var not in entry:
+            if -var in entry:
+                index = np.argwhere(entry==-var)
+                new_entry = np.delete(entry,index)
+                new_cnf.append(new_entry)
+            else:
+                new_cnf.append(entry)  
+    return new_cnf
+
+# find the most variables occouring most often across al clauses 
+def findMostCommonVar(cnf):
+    # create an array containing all the variables as often as they occour in the clauses (in either negated and non-negated state)
+    knf_variables_list = np.array([])
+    for array in cnf:
+        for entry in array:
+            knf_variables_list = np.append(knf_variables_list,abs(entry))
+    # count the occourance of the variables
+    knf_variables_counts = np.unique(knf_variables_list, return_counts=True)
+    # and choose the most frequent one
+    var = np.select(knf_variables_counts[1]==np.max(knf_variables_counts[1]),knf_variables_counts[0]).item()
+    # TODO: if we will not end up picking the variables in a different way anyways, it would make sense to also check weather the negated
+    # or the non-negated form are more common and pick that as the variable.
+    return var
+
+# since it can't simply be checked, weather an empty array is in a list (empty arrays are assigned a truth value by default), this function checks it
+def checkIfEmpty(cnf):
+    empty = False
+    for element in cnf:
+        if len(element)==0:
+            empty = True
+    return empty
+
+
+#######################################################################################################
+########### DPLL ITSELF ###############################################################################
+#######################################################################################################
+
+# we initialize an empty list of literals that will be filled with variable assignments in dpll 
+var = list()
+
+def dpll_alg(cnf):
+    # empty cnf is defined to be valid and therefore satisfiable
+    if cnf == []:
+        return list()
+    # cnf containing empty clauses are by convention unsatisfiable
+    elif checkIfEmpty(cnf):
+        return "unsat"
+    
+    # If neither of these cases occour, we start to search the tree for satisfying variable assignments.
+    # we do this recursively by repeatedly calling dpll_alg and assigning variables through conditioning
+    # until one of the first two cases applies and the cnf is therefore either sat or unsat.
+    else:
+        # we first select the variable for conditioning 
```

Comment:
here you could make it a little more readable if you substitute the nested "Else" statements into one larger "elif" statement imo.

Reply:
to explain further what i mean

```
if statement:
    ...
else:
    if statement2:
        ...
    else:
        ...
```

is equivalent to

```
if statement:
    ...
elif statement2:
    ...
else:
    ...
```

Reply:
That would work but I think I would still have to either of the two following things when it comes to the calculation of the new variable:
- define it at the beginning of the dpll_alg (i.e. before the if-else conditions) and therefore calculate the new variable (findMostCommonVar(cnf)) either every time dpll_alg is called even if one of the first two conditions is satisfied and it isn't necessary
- not define the new variable (or define it after already running the function that chooses it in order to determine weather we are in a case where we need to add it to the list of variables) but instead calculate it every time it is either needed in the condition or appended to the list of variables. But That way I would also run the operation multiple times instead of just calling the function that chooses the next variable one.

Or am I overseeing something? ;)

Reply:
Alo I thought that readability might benefit from defining the variables first as "new_cnf" and "var_candidate" rather than just using the functions that calculate them directly in the conditions. I figured this way it might be a bit easier to follow what the operations actually do but if you guys think that this is superfluous I can for sure avoid defining "new_cnf" by just using the function "conditioning" directly in the elif or if statement the way you proposed @Luc4sQ ! :)

Reply:
i've looked a little more deep into that and you're right. I overlooked this detail. that would end up in unnecessary calls of functions, as i don't know if python supports variable definition in a function call. (I think i saw that somewhere, but im not sure)

if you prefer your style in terms of readability, that is totally fine for me :) With my critic i was primarily aiming for sharing my opinion